contracts/src/KeyManager.sol

```diff
@@ -0,0 +1,224 @@
+            if (effectiveTimestamp <= lastTimestamp) {
+                revert InvalidEffectiveTimestamp(effectiveTimestamp, lastTimestamp);
+            }
+        }
+
+        if (committees.length > type(uint64).max) revert CommitteeIdOverflow();
+
+        committeeId = uint64(committees.length);
+        Committee storage newCommittee = committees.push();
+        newCommittee.effectiveTimestamp = effectiveTimestamp;
+        newCommittee.id = committeeId;
+        newCommittee.committeeMembers = members;
+
+        emit ScheduledCommittee(committeeId, effectiveTimestamp, uint64(members.length), keccak256(abi.encode(members)), msg.sender);
+        return committeeId;
+    }
+
+    /**
+     * @notice This function is used to get the committee by id.
+     * @dev Reverts if the id is greater than the length of the committees array.
+     * @dev Reverts if the committees array is empty.
+     * @dev Assumes that committees are not deleted from the committee array.
+     * @param id The id of the committee.
+     * @return effectiveTimestamp The effective timestamp of the committee.
+     * @return committeeMembers The committee members.
+     */
+    function getCommitteeById(uint64 id)
+        external
+        virtual
+        view
+        returns (uint64 effectiveTimestamp, CommitteeMember[] memory committeeMembers)
+    {
+        if (committees.length == 0) {
+            revert NoCommitteees();
+        }
+        if (id >= committees.length) {
+            revert CommitteeIdDoesNotExist(id, uint64(committees.length));
+        }
+        return (committees[id].effectiveTimestamp, committees[id].committeeMembers);
+    }
+
+    /**
+     * @notice This function is used to get the current committee id.
+     * @dev Reverts if the committees array is empty.
+     * @dev Reverts if there is no committee scheduled.
+     * @dev Assumes that committees are stored in ascending order of effective timestamp.
+     * @dev Assumes that committees are not deleted from the committee array.
+     * @return committeeId The current committee id.
+     */
+    function currentCommitteeId() public virtual view returns (uint64 committeeId) {
+        if (committees.length == 0) {
+            revert NoCommitteees();
+        }
+
+        uint64 currentTimestamp = uint64(block.timestamp);
+        if (currentTimestamp < committees[0].effectiveTimestamp) {
+            revert NoCommitteeScheduled(committees[0].effectiveTimestamp);
+        }
+        if (currentTimestamp >= committees[committees.length - 1].effectiveTimestamp) {
+            return committees[committees.length - 1].id;
+        }
+
+        uint256 lo = 0;
+        uint256 hi = committees.length - 1;
+        while (lo < hi) {
+            uint256 mid = (lo + hi + 1) / 2;
+            if (committees[mid].effectiveTimestamp <= currentTimestamp) {
+                lo = mid;
+            } else {
+                hi = mid - 1;
+            }
+        }
+        return committees[lo].id;
```

Comment:
imo, we don't need to do binary search, just reverse search is good. because it's either the last, or second last.

we can also avoid comparing with the first, simply start with the last, and find backwards.